File: ROT18.cpp
#include <iostream>
#include <regex>

using namespace std;

void cipherDecryption(){
    string rot5 = "5678901234";
    string zeroToNine = "0123456789";
    int rot13Key = 13;

    cout << "Message can be alphanumeric" << endl;
    string message;
    cout << "Enter message: ";
    getline(cin, message);
    cin.ignore();

    // message to uppercase
    for(int i = 0; i < message.length(); i++){
        message[i] = toupper(message[i]);
    }

    string decrypText = "";
    regex intReg("[\\s\\d]+");
    regex strReg("[\\s\\w]+");

    for (int i = 0; i < message.length(); i++){
        string temp = "";
        temp += message[i];

        if(message[i] == 32){
            decrypText += " ";
        } else if (regex_match(temp, intReg)){
            // ROT5
            for (int j = 0; j < zeroToNine.length(); j++){
                if (message[i] == rot5[j]){
                    decrypText += zeroToNine[j];
                }
            } // inner for
        } else if (regex_match(temp, strReg)){
            // ROT13
            int chTemp = temp[0] - rot13Key;
            if(chTemp < 65){
                chTemp += 26;
                decrypText += (char)chTemp;
            } else {
                decrypText += (char)chTemp;
            }
        } // if else
    } // for

    cout << "Decrypted Text: " << decrypText;
}

void cipherEncryption(){
    string rot5 = "5678901234";
    string zeroToNine = "0123456789";
    int rot13Key = 13;

    cout << "Message can be alphanumeric" << endl;
    string message;
    cout << "Enter message: ";
    getline(cin, message);
    cin.ignore();

    // message to uppercase
    for(int i = 0; i < message.length(); i++){
        message[i] = toupper(message[i]);
    }

    string encrypText = "";
    regex intReg("[\\s\\d]+");
    regex strReg("[\\s\\w]+");

    for (int i = 0; i < message.length(); i++){
        string temp = "";
        temp += message[i];

        if(message[i] == 32){
            encrypText += " ";
        } else if (regex_match(temp, intReg)){
            // ROT5
            for (int j = 0; j < zeroToNine.length(); j++){
                if (message[i] == zeroToNine[j]){
                    encrypText += rot5[j];
                }
            } // inner for
        } else if (regex_match(temp, strReg)){
            // ROT13
            int chTemp = temp[0] + rot13Key;
            if(chTemp > 90){
                chTemp -= 26;
                encrypText += (char)chTemp;
            } else {
                encrypText += (char)chTemp;
            }
        } // if else
    } // for

    cout << "Encrypted Text: " << encrypText;
}

int main()
{
    cout << "1. Encryption\n2. Decryption\n3. Choose(1,2): " << endl;
    int choice;
    cin >> choice;
    cin.ignore();

    if(choice == 1){
        cout << endl << "---Encryption---" << endl;
        cipherEncryption();

    } else if(choice == 2){
        cout << endl << "---Decryption---" << endl;
        cipherDecryption();

    } else {
        cout << endl << "Invalid Choice" << endl;
    }
    return 0;
}

File: ROT18.py
import re
def cipher_encryption():
    rot5 = "5678901234"
    zeroToNine = "0123456789"
    rot_13_key = 13

    print("Message can be alphanumeric")
    message = input("Enter message: ").upper()

    encryp_text = "";
    for i in range(len(message)):
        temp = message[i] + ""
        if ord(message[i]) == 32:
            encryp_text += " "
        elif re.search('[\d\s]+', temp):
            # ROT5
            for j in range(len(zeroToNine)):
                if message[i] == zeroToNine[j]:
                    encryp_text += rot5[j]
            # inner for
        elif re.search('[\w\s]+', temp):
            # ROT13
            ch_temp = ord(message[i]) + rot_13_key
            if ord(message[i]) == 32:
                encryp_text += " "
            elif ch_temp > 90:
                ch_temp -= 26
                encryp_text += chr(ch_temp)
            else:
                encryp_text += chr(ch_temp)
        # if-else
    # for

    print("Encrypted Text: {}".format(encryp_text))


def cipher_decryption():
    rot5 = "5678901234"
    zeroToNine = "0123456789"
    rot_13_key = 13

    print("Message can be alphanumeric")
    message = input("Enter message: ").upper()

    decryp_text = "";
    for i in range(len(message)):
        temp = message[i] + ""
        if ord(message[i]) == 32:
            decryp_text += " "
        elif re.search('[\d\s]+', temp):
            # ROT5
            for j in range(len(zeroToNine)):
                if message[i] == rot5[j]:
                    decryp_text += zeroToNine[j]
                    # inner for
        elif re.search('[\w\s]+', temp):
            # ROT13
            ch_temp = ord(message[i]) - rot_13_key
            if ord(message[i]) == 32:
                decryp_text += " "
            elif ch_temp < 65:
                ch_temp += 26
                decryp_text += chr(ch_temp)
            else:
                decryp_text += chr(ch_temp)
                # if-else
    # for

    print("Decrypted Text: {}".format(decryp_text))


def main():
    choice = int(input("1. Encryption\n2. Decryption\nChoose (1,2): "))
    if choice == 1:
        print("---Encryption---")
        cipher_encryption()
    elif choice == 2:
        print("---Decryption---")
        cipher_decryption()
    else:
        print("Invalid Choice")


if __name__ == "__main__":
    main()
